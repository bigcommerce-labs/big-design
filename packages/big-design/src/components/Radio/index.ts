export { Radio, type RadioProps } from './Radio';
export { RadioLabel } from './Label';
