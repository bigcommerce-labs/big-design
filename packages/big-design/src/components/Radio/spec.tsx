import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';

import { Radio } from './index';
import { StyleableRadio } from './private';

test('render Radio (checked)', () => {
  const { container } = render(<Radio label="Checked" name="test-group" checked={true} onChange={() => null} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('render Radio (unchecked)', () => {
  const { container } = render(<Radio label="Unchecked" name="test-group" checked={false} onChange={() => null} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('render Radio (checked + disabled)', () => {
  const { container } = render(
    <Radio label="Checked" name="test-group" checked={true} onChange={() => null} disabled />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('render Radio (unchecked + disabled)', () => {
  const { container } = render(
    <Radio label="Unchecked" name="test-group" checked={false} onChange={() => null} disabled />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('has correct value when checked', () => {
  const { getByTestId } = render(<Radio label="Checked" checked={true} onChange={() => null} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  expect(radio.checked).toBe(true);
});

test('has correct value when unchecked', () => {
  const { getByTestId } = render(<Radio label="Unchecked" checked={false} onChange={() => null} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  expect(radio.checked).toBe(false);
});

test('triggers onChange when clicking the radio button', () => {
  const onChange = jest.fn();
  const { getByTestId } = render(<Radio label="Checked" checked={false} onChange={onChange} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  fireEvent.click(radio);

  expect(onChange).toHaveBeenCalled();
});

test('triggers onChange when clicking styled and text label', () => {
  const onChange = jest.fn();
  const { container } = render(<Radio label="Checked" checked={false} onChange={onChange} />);

  const labels = container.querySelectorAll('label');

  labels.forEach(label => fireEvent.click(label));

  expect(onChange).toHaveBeenCalledTimes(2);
});

test('forwards ref', () => {
  const ref = React.createRef<HTMLInputElement>();

  const { container } = render(<Radio ref={ref} label="Checked" />);
  const input = container.querySelector('input');

  expect(input).toBe(ref.current);
});

test('does not forward styles', () => {
  const { container } = render(<Radio label="Checked" className="test" />);

  expect(container.getElementsByClassName('test').length).toBe(0);
});

test('private version forwards styles', () => {
  const { container } = render(<StyleableRadio label="Checked" className="test" />);

  expect(container.getElementsByClassName('test').length).toBe(1);
});
