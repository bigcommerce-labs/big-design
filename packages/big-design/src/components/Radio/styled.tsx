import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import { hideVisually } from 'polished';
import styled, { css } from 'styled-components';

import { withTransition } from '../../helpers/transitions';

interface StyledRadioProps {
  checked?: boolean;
  disabled?: boolean;
}

export const StyledRadioLabelContainer = styled.div`
  margin-left: ${({ theme }) => theme.spacing.xSmall};
`;

export const StyledRadioContainer = styled.div`
  align-items: flex-start;
  display: flex;
`;

export const StyledHiddenRadio = styled.input`
  ${hideVisually()}
`;

export const StyledRadio = styled.label<StyledRadioProps>`
  ${withTransition(['border-color', 'box-shadow'])}

  background-color: ${({ disabled, theme }) =>
    disabled ? theme.colors.secondary10 : theme.colors.white};
  border: ${({ theme }) => theme.border.box};
  border-color: ${(props) =>
    props.checked ? props.theme.colors.primary40 : props.theme.colors.secondary30};
  border-radius: ${({ theme }) => theme.borderRadius.circle};
  color: ${({ theme }) => theme.colors.white};
  cursor: pointer;
  display: inline-block;
  flex-shrink: 0;
  height: ${({ theme }) => theme.spacing.large};
  position: relative;
  user-select: none;
  width: ${({ theme }) => theme.spacing.large};

  ${({ disabled, theme }) =>
    disabled &&
    css`
      cursor: not-allowed;
      border-color: ${theme.colors.secondary30};
    `};

  &:hover {
    border-color: ${({ checked, disabled, theme }) =>
      !disabled && (checked ? theme.colors.primary50 : theme.colors.secondary40)};
  }

  ${StyledHiddenRadio}:focus + & {
    box-shadow: ${({ theme }) => `0 0 0 ${theme.spacing.xxSmall} ${theme.colors.primary20}`};
  }

  &:after {
    ${withTransition(['opacity'])}

    background-color: ${({ disabled, theme }) =>
      disabled ? theme.colors.secondary40 : theme.colors.primary40};
    border-radius: 50%;
    content: '';
    height: ${({ theme }) => theme.spacing.small};
    left: 50%;
    opacity: 0;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%);
    width: ${({ theme }) => theme.spacing.small};
  }

  ${(props) =>
    props.checked &&
    css`
      &:after {
        opacity: 1;
      }
    `}
`;

StyledRadio.defaultProps = { theme: defaultTheme };
StyledRadioLabelContainer.defaultProps = { theme: defaultTheme };
StyledRadioContainer.defaultProps = { theme: defaultTheme };
StyledHiddenRadio.defaultProps = { theme: defaultTheme };
