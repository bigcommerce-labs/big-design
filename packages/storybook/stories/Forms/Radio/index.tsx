export * from './Radio';
export * from './RadioPropTable';
