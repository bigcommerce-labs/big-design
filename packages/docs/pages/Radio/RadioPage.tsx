import { Fieldset, Form, H0, H1, Link, Radio, Text } from '@bigcommerce/big-design';
import React from 'react';

import { Code, CodePreview } from '../../components';
import { RadioPropTable } from '../../PropTables';

export default () => (
  <>
    <H0>Radio</H0>

    <Text>
      Radios are single-selectable form control groups.{' '}
      <Link href="https://design.bigcommerce.com/components/radio-buttons" target="_blank">
        Radio Design Guidelines
      </Link>
      .
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [selected, setSelected] = React.useState('');

        const handleChange = event => setSelected(event.target.value);

        return (
          <Form>
            <Form.Group>
              <Radio label="On" checked={selected === 'on'} value="on" onChange={handleChange} />
              <Radio label="Off" checked={selected === 'off'} value="off" onChange={handleChange} />
            </Form.Group>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>

    <H1>API</H1>

    <Text>
      Supports all native <Code>&lt;input /&gt;</Code> element attributes.
    </Text>

    <RadioPropTable />

    <H1>Grouping</H1>

    <Text>
      In order to group radio controls, use the <Code>Fieldset</Code> component to separate the controls.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [firstRadio, setFirstRadio] = React.useState('');
        const [secondRadio, setSecondRadio] = React.useState('');

        const handleFirstChange = event => setFirstRadio(event.target.value);
        const handleSecondChange = event => setSecondRadio(event.target.value);

        return (
          <Form>
            <Fieldset legend="First Group">
              <Form.Group>
                <Radio label="On" checked={firstRadio === 'on'} value="on" onChange={handleFirstChange} />
                <Radio label="Off" checked={firstRadio === 'off'} value="off" onChange={handleFirstChange} />
              </Form.Group>
            </Fieldset>
            <Fieldset legend="Second Group">
              <Form.Group>
                <Radio label="On" checked={secondRadio === 'on'} value="on" onChange={handleSecondChange} />
                <Radio label="Off" checked={secondRadio === 'off'} value="off" onChange={handleSecondChange} />
              </Form.Group>
            </Fieldset>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>
  </>
);
