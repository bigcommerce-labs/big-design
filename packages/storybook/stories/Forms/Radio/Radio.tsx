import { Form, H0, H1, H2, Link, Radio, Text } from '@bigcommerce/big-design';
import React from 'react';

import { Code, CodePreview } from '../../../components';

import { RadioPropTable } from './index';

export const RadioStory: React.FC = () => (
  <>
    <H0>Radio</H0>

    <Text>
      Radios are single-selectable form control groups.{' '}
      <Link href="https://bigcommerce.design/fields" target="_blank">
        Form Fields Design Guidelines
      </Link>
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [selected, setSelected] = React.useState('');

        const handleChange = event => setSelected(event.target.value);

        return (
          <Form>
            <Form.Row>
              <Radio label="On" checked={selected === 'on'} value="on" onChange={handleChange} />
            </Form.Row>
            <Form.Row>
              <Radio label="Off" checked={selected === 'off'} value="off" onChange={handleChange} />
            </Form.Row>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>

    <H1>API</H1>

    <H2>Radio</H2>

    <Text>
      Supports all native <Code>&lt;input /&gt;</Code> element attributes.
    </Text>

    <RadioPropTable />

    <H1>Grouping</H1>

    <Text>
      In order to group radio controls, use the <Code>Form.Fieldset</Code> component to separate the controls.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [firstRadio, setFirstRadio] = React.useState('');
        const [secondRadio, setSecondRadio] = React.useState('');

        const handleFirstChange = event => setFirstRadio(event.target.value);
        const handleSecondChange = event => setSecondRadio(event.target.value);

        return (
          <Form>
            <Form.Fieldset legend="First Group">
              <Form.Row>
                <Radio label="On" checked={firstRadio === 'on'} value="on" onChange={handleFirstChange} />
              </Form.Row>
              <Form.Row>
                <Radio label="Off" checked={firstRadio === 'off'} value="off" onChange={handleFirstChange} />
              </Form.Row>
            </Form.Fieldset>
            <Form.Fieldset legend="Second Group">
              <Form.Row>
                <Radio label="On" checked={secondRadio === 'on'} value="on" onChange={handleSecondChange} />
              </Form.Row>
              <Form.Row>
                <Radio label="Off" checked={secondRadio === 'off'} value="off" onChange={handleSecondChange} />
              </Form.Row>
            </Form.Fieldset>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>
  </>
);
