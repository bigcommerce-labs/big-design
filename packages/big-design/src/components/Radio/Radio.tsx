import React, { forwardRef, useMemo, Ref } from 'react';

import { useUniqueId } from '../../hooks';
import { typedMemo, warning } from '../../utils';

import { HiddenRadio, RadioContainer, StyledRadio } from './styled';
import { RadioLabel } from './Label';

interface Props {
  label: React.ReactChild;
}

interface PrivateProps {
  forwardedRef: Ref<HTMLInputElement>;
}

export type RadioProps = Props & React.InputHTMLAttributes<HTMLInputElement>;

const RawRadio: React.FC<RadioProps & PrivateProps> = ({
  checked,
  className,
  disabled,
  label,
  forwardedRef,
  style,
  ...props
}) => {
  const uniqueRadioId = useUniqueId('radio');
  const id = props.id ? props.id : uniqueRadioId;
  const labelId = useUniqueId('radio_label');

  const renderedLabel = useMemo(() => {
    if (!label) {
      return null;
    }

    if (typeof label === 'string') {
      return (
        <RadioLabel htmlFor={id} id={labelId} disabled={disabled} aria-hidden={disabled}>
          {label}
        </RadioLabel>
      );
    }

    if (React.isValidElement(label) && label.type === RadioLabel) {
      return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
        htmlFor: id,
        id: labelId,
      });
    }

    warning('label must be either a string or a RadioLabel component.');
  }, [disabled, id, label, labelId]);

  return (
    <RadioContainer className={className} style={style}>
      <HiddenRadio
        type="radio"
        checked={checked}
        id={id}
        disabled={disabled}
        {...props}
        aria-labelledby={labelId}
        ref={forwardedRef}
      />
      <StyledRadio checked={checked} disabled={disabled} htmlFor={id} aria-hidden={true} />
      {renderedLabel}
    </RadioContainer>
  );
};

export const Radio = typedMemo(
  forwardRef<HTMLInputElement, RadioProps>(({ className, style, ...props }, ref) => (
    <RawRadio {...props} forwardedRef={ref} />
  )),
);
