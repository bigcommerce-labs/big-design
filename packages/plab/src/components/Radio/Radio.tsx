import React from 'react';

import { uniqueId } from '../../utils';

import { HiddenRadio, RadioContainer, StyledLabel, StyledRadio } from './styled';

interface Props {
  label: React.ReactChild;
}

export type RadioProps = Props & React.InputHTMLAttributes<HTMLInputElement>;

export class Radio extends React.PureComponent<RadioProps> {
  static Label = StyledLabel;
  private readonly uniqueId = uniqueId('radio_');

  render() {
    const { checked, className, label, ...props } = this.props;
    const id = this.getId();

    return (
      <RadioContainer className={className}>
        <HiddenRadio type="radio" checked={checked} id={id} {...props} />
        <StyledRadio checked={checked} htmlFor={id} aria-hidden={true} />
        {this.renderLabel()}
      </RadioContainer>
    );
  }

  private getId() {
    const { id } = this.props;

    return id ? id : this.uniqueId;
  }

  private renderLabel() {
    const id = this.getId();
    const { label } = this.props;

    if (typeof label === 'string') {
      return <StyledLabel htmlFor={id}>{label}</StyledLabel>;
    }

    if (React.isValidElement(label) && label.type === Radio.Label) {
      return React.cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, {
        htmlFor: id,
      });
    }

    return null;
  }
}
