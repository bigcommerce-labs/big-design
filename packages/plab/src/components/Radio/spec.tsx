import 'jest-styled-components';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';

import { Radio } from './Radio';

test('render Radio (checked)', () => {
  const { container } = render(<Radio label="Checked" name="test-group" checked={true} onChange={() => null} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('render Radio (unchecked)', () => {
  const { container } = render(<Radio label="Unchecked" name="test-group" checked={false} onChange={() => null} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('has correct value when checked', () => {
  const { getByTestId } = render(<Radio label="Checked" checked={true} onChange={() => null} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  expect(radio.checked).toBe(true);
});

test('has correct value when unchecked', () => {
  const { getByTestId } = render(<Radio label="Unchecked" checked={false} onChange={() => null} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  expect(radio.checked).toBe(false);
});

test('triggers onChange when clicking the radio button', () => {
  const onChange = jest.fn();
  const { getByTestId } = render(<Radio label="Checked" checked={false} onChange={onChange} data-testid="radio" />);
  const radio = getByTestId('radio') as HTMLInputElement;

  fireEvent.click(radio);

  expect(onChange).toHaveBeenCalled();
});

test('triggers onChange when clicking styled and text label', () => {
  const onChange = jest.fn();
  const { container } = render(<Radio label="Checked" checked={false} onChange={onChange} />);

  const labels = container.querySelectorAll('label');

  labels.forEach(label => fireEvent.click(label));

  expect(onChange.mock.calls.length).toBe(2);
});
