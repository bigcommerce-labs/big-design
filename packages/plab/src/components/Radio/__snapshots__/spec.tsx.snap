// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Radio (checked) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c4 {
  margin-left: 1rem;
}

.c3 {
  border-radius: 50%;
  border: 1px solid #D9DCE9;
  border-color: #3C64F4;
  color: #FFFFFF;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1.25rem;
}

.c1:focus + .c3 {
  box-shadow: 0 0 0 4px #DBE3FE;
}

.c3:after {
  background-color: #3C64F4;
  border-radius: 50%;
  content: '';
  height: 0.75rem;
  left: 50%;
  position: absolute;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  width: 0.75rem;
}

<div
  class="c0"
>
  <input
    aria-labelledby="checkBox_label_2"
    checked=""
    class="c1 c2"
    id="radio_1"
    name="test-group"
    type="radio"
  />
  <label
    aria-hidden="true"
    class="c3"
    for="radio_1"
  />
  <label
    class="c4"
    for="radio_1"
    id="checkBox_label_2"
  >
    Checked
  </label>
</div>
`;

exports[`render Radio (unchecked) 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c4 {
  margin-left: 1rem;
}

.c1:focus + .c5 {
  box-shadow: 0 0 0 4px #DBE3FE;
}

.c3 {
  border-radius: 50%;
  border: 1px solid #D9DCE9;
  border-color: #D9DCE9;
  color: #FFFFFF;
  display: inline-block;
  height: 1.25rem;
  position: relative;
  -webkit-transition: all 150ms;
  transition: all 150ms;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1.25rem;
}

.c1:focus + .c3 {
  box-shadow: 0 0 0 4px #DBE3FE;
}

<div
  class="c0"
>
  <input
    aria-labelledby="checkBox_label_4"
    class="c1 c2"
    id="radio_3"
    name="test-group"
    type="radio"
  />
  <label
    aria-hidden="true"
    class="c3"
    for="radio_3"
  />
  <label
    class="c4"
    for="radio_3"
    id="checkBox_label_4"
  >
    Unchecked
  </label>
</div>
`;
